### PR TITLE
Refactor Clockface Form and Grid

### LINKED ISSUE #2045: Add clockface input validation
A user should be able to type into a clockface form input that requires validations and get feedback from error messages

### PR DESCRIPTION
**Pull request by alexpaxton**
Closes #2045 

form-validation-2

### Changes

- Separate concerns of `Form` components into `Grid` (for layout) and `Form` for building forms
- Anywhere `Form` was previously used now uses both `Form` and `Grid`
- `Grid` can be used anywhere
- `Form` now offers `Form.ValidationElement` (See below)

### Introducing `<Form.ValidationElement />`

Receives the same value as its child input in addition to a validation function. Depending on whether the validation function returns an error message or not the `Validation.Element` passes a `ComponentStatus` to its child via render prop.

```
import {Form, ComponentStatus, Input} from 'src/clockface'
```
```
this.state = {
  inputValue: '',
}
```
```
const {inputValue} = this.state

<Form.ValidationElement
  label="Enter a Hexcode"
  value={inputValue}
  validationFunc={this.validateColorHex}
>
  {status => (
    <Input
      value={inputValue}
      placeholder="#000000"
      onChange={this.handleInputChange}
      status={status}
      maxLength={7}
    />
  )}
</Form.ValidationElement>
```
```
private validateColorHex = (colorHex): string | null => {
  const isValidLength = colorHex.length === 7
  const containsValidCharacters =
    colorHex.replace(/[ABCDEF0abcdef123456789]+/g, '') === '#'

  const errorMessage = []

  if (!containsValidCharacters) {
    errorMessage.push('Hexcodes must begin with # and include A-F 0-9')
  }

  if (!isValidLength) {
    errorMessage.push('Hexcodes must be 7 characters')
  }

  return errorMessage.join(', ')
}
```

### Introducing `<Grid />`

This component set allows anything to be laid out in a 12 column grid. There were previously a set of CSS classes for handling this, but without being aware of all the classes it is not that useful.

```
import {Grid, Columns} from 'src/clockface'
```
```
<Grid>
  <Grid.Row>
    <Grid.Column widthXS={Columns.Twelve} widthSM={Columns.Six} widthMD={Columns.Four} widthLG={Columns.Three}>
      [component goes here]
    </Grid.Column>
  </Grid.Row>
</Grid>
```

There are a number of instances of plain HTML that offers the same layout features that should be replaced with `Grid`.

### Checklist

  - [x] Rebased/mergeable
  - [x] Tests pass

### CODE REVIEW

File: ui/src/onboarding/components/AdminStep.tsx [hunk 1]
```diff
@@ -68,82 +69,80 @@ class AdminStep extends PureComponent<OnboardingStepProps, State> {
...
+                    icon={icon}
+                    status={status}
+                    disabledTitleText="Admin username has been set"
+                    autoFocus={true}
+                  />
+                </Form.Element>
+              </Grid.Column>
+              <Grid.Column widthXS={Columns.Three} offsetXS={Columns.Three}>
+                <Form.Element label="Admin Password">
+                  <Input
```
Comment (ischolten) [inline ui/src/onboarding/components/AdminStep.tsx]: should the password and confirm password be in the same row as the username?

File: ui/src/clockface/components/grid_layout/Grid.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import {ErrorHandling} from 'src/shared/decorators/errors'
+
+interface Props {
+  children: JSX.Element[] | JSX.Element
+}
+
+@ErrorHandling
+class Grid extends Component<Props> {
+  public static Row = GridRow
+  public static Column = GridColumn
```
Comment (ischolten) [inline ui/src/clockface/components/grid_layout/Grid.tsx]: where is column used?

File: ui/src/me/components/account/Settings.tsx [hunk 3]
```diff
@@ -40,16 +41,22 @@ export class Settings extends PureComponent<StateProps, State> {
...
-              <Input
-                value={me.name}
-                dataTest="nameInput"
-                titleText="Username"
-                size={ComponentSize.Small}
-                status={ComponentStatus.Disabled}
-                onChange={this.handleChangeInput}
-              />
-            </Form.Element>
+            <Grid>
```
Comment (ischolten) [inline ui/src/me/components/account/Settings.tsx]: will all forms require the use of grid component now?

File: ui/src/onboarding/components/AdminStep.tsx [hunk 4]
```diff
@@ -68,82 +69,80 @@ class AdminStep extends PureComponent<OnboardingStepProps, State> {
...
-              titleText="Default Bucket Name"
-              size={ComponentSize.Medium}
-              icon={icon}
-              status={status}
-              placeholder="Your bucket is where you will store all your data"
-              disabledTitleText="Default bucket name has been set"
-            />
-          </Form.Element>
+          <Grid>
+            <Grid.Row>
```
Comment (ischolten) [inline ui/src/onboarding/components/AdminStep.tsx]: to clarify, this Grid.Row will move the columns down if theres not space for them based on the width specified by props?

File: ui/src/onboarding/components/AdminStep.tsx [hunk 5]
```diff
@@ -68,82 +69,80 @@ class AdminStep extends PureComponent<OnboardingStepProps, State> {
...
+                    titleText="Admin Username"
+                    size={ComponentSize.Medium}
+                    icon={icon}
+                    status={status}
+                    disabledTitleText="Admin username has been set"
+                    autoFocus={true}
+                  />
+                </Form.Element>
+              </Grid.Column>
+              <Grid.Column widthXS={Columns.Three} offsetXS={Columns.Three}>
```
Comment (ischolten) [inline ui/src/onboarding/components/AdminStep.tsx]: could additional rows be used here to force new rows? like keeping the username in one row, the password and confirm password in another, and the org in a final one?

File: ui/src/shared/components/view_options/options/TimeFormat.tsx [hunk 6]
```diff
@@ -33,52 +33,62 @@ class TimeFormat extends PureComponent<Props, State> {
...
-            {showCustom && (
-              <a href={TIME_FORMAT_TOOLTIP_LINK} target="_blank">
-                <QuestionMarkTooltip
-                  tipID="Time Axis Format"
-                  tipContent={tipContent}
-                />
-              </a>
+      <Grid.Column widthXS={Columns.Twelve}>
+        <Form.Element label="Time Format" labelAddOn={this.timeFormatTooltip}>
+          <>
```
Comment (ischolten) [inline ui/src/shared/components/view_options/options/TimeFormat.tsx]: is this fragment necessary?

Comment (alexpaxton) [reply ui/src/shared/components/view_options/options/TimeFormat.tsx]: Yarrr Form elements expect one child, which exists as a friendly nod to devs to keep it simple

File: ui/src/clockface/components/grid_layout/Grid.scss [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+@import 'src/style/modules';
+
+$grid--gutter: 8px;
+
+$grid--breakpoint-sm: 750px;
+$grid--breakpoint-md: 992px;
+$grid--breakpoint-lg: 1200px;
+
+$grid--col-1: 8.33333333%;
```
Comment (mavarius) [inline ui/src/clockface/components/grid_layout/Grid.scss]: Could you use the sass percentage function for this?

**Pull request merged**